app/Models/Beatmapset.php

```diff
@@ -601,29 +601,34 @@ public function regenerateCovers()
 
             // download and extract beatmap
             $osz = "$tmpBase/osz.zip";
-            $ok = copy($this->oszDownloadURL(), $osz);
-            if (!$ok) {
+            try {
+                $ok = copy($this->oszDownloadURL(), $osz);
+                if (!$ok) {
+                    throw new BeatmapProcessorException('Error retrieving beatmap');
+                }
+            } catch (\Exception $e) {
                 throw new BeatmapProcessorException('Error retrieving beatmap');
             }
+
             $zip = new \ZipArchive;
             $zip->open($osz);
             $zip->extractTo($workingFolder);
             $zip->close();
 
-            // grab the first beatmap (as per old implementation) and scan for background image
-            $beatmap = $this->beatmaps()->first();
-            $beatmapFilename = $beatmap->filename;
-            $bgFilename = self::scanBMForBG("$workingFolder/$beatmapFilename");
+            // scan through all the beatmaps in this set and find the first one with a valid background image
+            foreach ($this->beatmaps as $beatmap) {
+                $bgFilename = self::scanBMForBG("{$workingFolder}/{$beatmap->filename}");
 
-            if (!$bgFilename) {
-                $this->update(['cover_updated_at' => $this->freshTimestamp()]);
+                if ($bgFilename === false) {
```

Comment:
and then there might be case when the beatmap file exists but no background defined?